sync/pooled_worker_pool.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package sync implements synchronization facililites such as worker pools.
+package sync
+
+import "math/rand"
+
+const (
+	defaultNumShards = 10
+)
+
+var (
+	// Var for testing purposes
+	defaultKillWorkerProbability = 0.0001
+)
+
+// PooledWorkerPool provides a pool for goroutines, but unlike WorkerPool,
+// the actual goroutines themselves are re-used. This can be useful from a
+// performance perspective in scenarios where the allocation and growth of
+// the new goroutine and its stack is a bottleneck. Specifically, if the
+// work function being performed has a very deep call-stack, calls to
+// runtime.morestack can dominate the workload. Re-using existing goroutines
+// allows the stack to be grown once, and then re-used for many invocations.
+//
+// In order to prevent abnormally large goroutine stacks from persisting over
+// the life-cycle of an application, the PooledWorkerPool will randomly kill
+// existing goroutines and spawn a new one.
+//
+// The PooledWorkerPool also implements sharding of its underlying worker channels
+// to prevent excessive lock contention.
+type PooledWorkerPool interface {
+	// Init initializes the pool.
+	Init()
+
+	// Go waits until the next worker becomes available and executes it.
+	Go(work Work)
+}
+
+type pooledWorkerPool struct {
+	workChs   []chan workerInstruction
+	numShards int
+}
+
+type workerInstruction struct {
+	work      Work
+	shouldDie bool
+}
+
+// NewPooledWorkerPool creates a new worker pool.
+func NewPooledWorkerPool(size int) PooledWorkerPool {
```

Comment:
Nit: Add option for killProbability. Would be interesting to run in prod with a few different settings to prove out the effects in prod.

Reply:
made options